File: apps/api/src/devices/naming-tokens.ts
export interface LocationTokens { site: string; building: string; floor: string; area: string; role: string; }

const SEQ = '{seq}';

/** Substitute every token except {seq}; collapse separator runs left by empty tokens; trim edges. */
export function renderTemplate(template: string, tokens: LocationTokens): string {
  let out = template
    .replace(/\{site\}/g, tokens.site)
    .replace(/\{building\}/g, tokens.building)
    .replace(/\{floor\}/g, tokens.floor)
    .replace(/\{area\}/g, tokens.area)
    .replace(/\{role\}/g, tokens.role);
  // Protect {seq} from separator-collapsing.
  const guarded = out.split(SEQ);
  const cleaned = guarded.map((seg) => seg.replace(/[-_]{2,}/g, (m) => m[0]));
  out = cleaned.join(SEQ).replace(/^[-_]+/, '').replace(/[-_]+$/, '');
  return out;
}

export function hasSeqToken(rendered: string): boolean { return rendered.includes(SEQ); }
export function fillSeq(rendered: string, seq: string): string { return rendered.replace(SEQ, seq); }

File: src/Modules/Inventory/NodeScope.Modules.Inventory.Domain/NamingTokens.cs
using System.Text.RegularExpressions;

namespace NodeScope.Modules.Inventory.Domain;

/// <summary>Location tokens for device-name templates (Node's <c>naming-tokens.ts</c>).</summary>
public sealed record LocationTokens(string Site, string Building, string Floor, string Area, string Role);

/// <summary>Template rendering for name suggestions.</summary>
public static partial class NamingTokens
{
    public const string SeqToken = "{seq}";

    /// <summary>
    /// Substitutes every token except <c>{seq}</c>, collapses separator runs left by empty
    /// tokens, and trims separator edges.
    /// </summary>
    public static string RenderTemplate(string template, LocationTokens tokens)
    {
        ArgumentNullException.ThrowIfNull(template);
        ArgumentNullException.ThrowIfNull(tokens);
        var substituted = template
            .Replace("{site}", tokens.Site, StringComparison.Ordinal)
            .Replace("{building}", tokens.Building, StringComparison.Ordinal)
            .Replace("{floor}", tokens.Floor, StringComparison.Ordinal)
            .Replace("{area}", tokens.Area, StringComparison.Ordinal)
            .Replace("{role}", tokens.Role, StringComparison.Ordinal);

        // Protect {seq} from separator-collapsing.
        var segments = substituted.Split(SeqToken);
        var cleaned = segments.Select(segment => SeparatorRuns().Replace(segment, match => match.Value[..1]));
        return string.Join(SeqToken, cleaned).Trim('-', '_');
    }

    public static bool HasSeqToken(string rendered)
    {
        ArgumentNullException.ThrowIfNull(rendered);
        return rendered.Contains(SeqToken, StringComparison.Ordinal);
    }

    public static string FillSeq(string rendered, string seq)
    {
        ArgumentNullException.ThrowIfNull(rendered);
        var index = rendered.IndexOf(SeqToken, StringComparison.Ordinal);
        return index < 0 ? rendered : rendered[..index] + seq + rendered[(index + SeqToken.Length)..];
    }

    [GeneratedRegex("[-_]{2,}")]
    private static partial Regex SeparatorRuns();
}
